### ISSUE TITLE
Replace `instanceof Array` with `Array.isArray()`

### ISSUE THREAD

**Issue by fluidsonic** [angular/angular#32950]
# 🐞 bug report

### Affected Package
Looking at the source code at least `compiler`, `core`, and `forms`.

### Is this a regression?

I don't think so.

### Description

Angular code is using `instanceof Array` in several occasions which is `false` for array literals in a [Node VM context](https://nodejs.org/api/vm.html#vm_vm_executing_javascript). There `[] instanceof Array` will be `false`. That causes erratic behavior in Angular Universal/SSR when run in such a VM. One example in dependency injection (see exception below).

In https://github.com/nodejs/node/issues/3425 and https://github.com/nodejs/node/issues/11593 Node developers recommend to always use `Array.isArray()` instead of `instanceof Array`.

## 🔬 Minimal Reproduction

```sh
git clone git@github.com:fluidsonic/angular-instanceof-array.git
cd angular-instanceof-array
npm install
npm run build:ssr && npm run serve:ssr
```

## 🔥 Exception

Exception during dependency injection due to an array not being recognized as such. It will be treated as an object instead.

```
Error: StaticInjectorError[{0:[object Object], 1:[object Object], 2:[object Object], 3:[object Object], 4:[object Object], 5:[object Object], 6:[object Object], 7:[object Object], 8:[object Object], 9:[object Object], 10:[object Object], 11:[object Object],[object Object],[object Object],[object Object], 12:[object Object], 13:[object Object]}]: Unexpected provider
    at staticError (evalmachine.<anonymous>:8:416065)
    at recursivelyProcessProviders (evalmachine.<anonymous>:8:413706)
    at new StaticInjector (evalmachine.<anonymous>:8:414979)
    at INJECTOR_IMPL__PRE_R3__ (evalmachine.<anonymous>:8:412751)
    at Function.create (evalmachine.<anonymous>:8:412847)
    at evalmachine.<anonymous>:8:558017
    at evalmachine.<anonymous>:8:557889
    at _getPlatform (evalmachine.<anonymous>:8:1432596)
    at Module.renderModuleFactory (evalmachine.<anonymous>:8:1433567)
```

## 🌍  Your Environment

**Angular Version:**
```
Angular CLI: 8.3.6
Node: 12.10.0
OS: darwin x64
Angular: 8.2.8
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router

Package                                    Version
--------------------------------------------------------------------
@angular-devkit/architect                  0.803.6
@angular-devkit/build-angular              0.803.6
@angular-devkit/build-optimizer            0.803.6
@angular-devkit/build-webpack              0.803.6
@angular-devkit/core                       8.3.6
@angular-devkit/schematics                 8.3.6
@angular/cli                               8.3.6
@ngtools/webpack                           8.3.6
@nguniversal/common                        8.1.1
@nguniversal/module-map-ngfactory-loader   8.1.1
@schematics/angular                        8.3.6
@schematics/update                         0.803.6
rxjs                                       7.0.0-alpha.0
typescript                                 3.5.3
webpack                                    4.41.0
```

**Anything else relevant?**
This likely affects only Angular Universal/SSR.
Arrays passed between iframes in a browser however can cause the same problems.

**Comment by mhevery** [comment]
Would you be willing to do a PR to fix this?

**Comment by Potapy4** [comment]
@mhevery I can jump on it if @fluidsonic doesn't mind 🙂

**Comment by JoostK** [comment]
Closing as resolved in #33076, #33077 and #33078.
